pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/AuthenticationRoute.java

```diff
@@ -30,72 +35,158 @@
-                } else {
-                    String userSetting = gson.toJson(userUtil.getUserAccessRoles(email), ArrayList.class);
-                    claims.put(userAccessRoles, userSetting);
-                    logger.info(userSetting);
-                    claims.put(userSettings, gson.toJson(UserSettings.getUserSettings(email), UserSettings.class));
+        try {
+            JsonObject jsonObject = new JsonParser().parse(request.body()).getAsJsonObject();
+            String auth0Token = jsonObject.get(payloadToken).getAsString();
+            if (StringUtils.isNotBlank(auth0Token)) {
+                // checking if Auth0 knows that token
+                try {
+                    Auth0Util.Auth0UserInfo auth0UserInfo = auth0Util.getAuth0UserInfo(auth0Token, auth0Domain);
+                    if (auth0UserInfo != null) {
+                        String email = auth0UserInfo.getEmail();
+                        logger.info("User (" + email + ") was found ");
+                        Gson gson = new Gson();
+                        Map<String, String> claims = new HashMap<>();
+                        UserDao userDao = new UserDao();
+                        UserDto userDto = userDao.getUserByEmail(email).orElseThrow(() -> new RuntimeException("User " + email + " not found!"));
+                        if (userDto == null) {
+                            userUtil.insertUser(email, email);
+                            userDto = userDao.getUserByEmail(email).orElseThrow(() -> new RuntimeException("new inserted user " + email + " not found!"));
+                            claims.put(userAccessRoles, "user needs roles and groups");
+                        }
+                        else {
+                            String userSetting = gson.toJson(userUtil.getUserAccessRoles(email), ArrayList.class);
+                            claims.put(userAccessRoles, userSetting);
+                            logger.info(userSetting);
+                            claims.put(userSettings, gson.toJson(UserSettings.getUserSettings(email), UserSettings.class));
+                        }
+                        claims.put(authUserId, String.valueOf(userDto.getId()));
+                        claims.put(authUserName, userDto.getName().orElse(""));
+                        claims.put(authUserEmail, email);
+                        claims = getDSSClaimsFromOriginalToken(auth0Token, auth0Domain, claims);
+
+                        try {
+                            String dsmToken = getNewAuth0TokenWithCustomClaims(claims, clientSecret, auth0ClientId, auth0Domain, auth0MgmntAudience, audienceNameSpace);
+                            if (dsmToken != null) {
+                                return new DSMToken(dsmToken);
+                            }
+                            else {
+                                response.status(401);
+                                throw new RuntimeException("DSMToken was null! Not authorized user");
```

Comment:
Unless Spark does something surprising, I don't expect that your status codes will be sent to the client.
Setting the code in response is always followed by RuntimeException, so doubt Spark will get a chance to send the response you intended. More likely, client will get a 500/System error.

Reply:
I am now using `halt` like how DSS does it, I think this should work now 
(Also taking  note to make sure to write tests for all of these cases)